beps/0005-definitive-plugin-listing/README.md

```diff
@@ -166,21 +166,18 @@ interface BackstageInstance {
   };
 }
 
-interface DeploymentMetadataPluginExtensionPoint {
-  setInstances: (instances: BackstageInstance[]) => Promise<void>;
+interface DeploymentMetadataService {
+  listInstances: () => BackstageInstance[];
```

Comment:
hmm, wonder if this should just be the same list of features rather than a list of instances. A slight issue I see with this is that we don't have the concept of routing to an instance, only plugins. It'd be an additional thing to add to the system.

Reply:
Yeah, I call that out as a TODO below in the doc. I'll make that a bit more clear in the BEP - here's the original thread with Freben around plugins vs instances, https://github.com/backstage/backstage/pull/25431#discussion_r1722788174.

IMO we should stick with instances here and rely on the instance metadata service HTTP API to retrieve all metadata, the plugin-first approach is more discovery specific and in my mind is the custom discovery implementation using these 2 APIs. The big issue with just scraping the existing static config + list of the current instance's plugins is that it requires the config to be up to date with everything that's deployed. Couple of examples where I think that is a little rough,
1. Let's say you deploy the catalog + scaffolder to their own pods, but you have a single instance running smaller plugins, ie api-docs, techdocs, user-settings, etc. That single instance will have the correct list of plugins for the entire system, but unless you add each pluginId to your configs for catalog + scaffolder, they'd be inaccurate.
2. Similarly, if you want to dynamically add a plugin to a deployment, assuming you're just using the default discovery implementation, you'd need to adjust the config as well.

I think having the system metadata service provide instances and let the clients scrape instances themselves provides a bridge between the purely static and a more dynamic system that the future may hold.